crates/sui-graphql-rpc/src/types/transaction_block/mod.rs

```diff
@@ -315,17 +315,40 @@ impl TransactionBlock {
             }
         }
 
+        let cursor_viewed_at = page.validate_cursor_consistency()?;
+        let checkpoint_viewed_at = cursor_viewed_at.unwrap_or(checkpoint_viewed_at);
+        let db: &Db = ctx.data_unchecked();
+        let is_from_front = page.is_from_front();
+        // If we've entered this function, we already fetched `checkpoint_viewed_at` from the
+        // `Watermark`, and so we must be able to retrieve `lo_cp` as well.
+        let Watermark { lo_cp, lo_tx, .. } = *ctx.data_unchecked();
+        let cp_after = filter.after_checkpoint.map(u64::from);
+        let cp_at = filter.at_checkpoint.map(u64::from);
+        let cp_before = filter.before_checkpoint.map(u64::from);
+
+        // Check that requested data is not already pruned. This should catch most cases, except on
+        // epoch boundary where a request is made just before the epoch pruning process is kicked
+        // off.
+        if checkpoint_viewed_at < lo_cp
+            || cp_after.map_or(false, |cp| cp < lo_cp)
+            || cp_at.map_or(false, |cp| cp < lo_cp)
+            || cp_before.map_or(false, |cp| cp < lo_cp)
+            || page.after().map_or(false, |c| c.tx_sequence_number < lo_tx)
+            || page
+                .before()
+                .map_or(false, |c| c.tx_sequence_number < lo_tx)
+        {
+            return Err(Error::Client(
+                "Requested data is pruned and no longer available".to_string(),
+            ));
```

Comment:
I think @lxfind mentioned that it would be better to treat pruned data the same as if it had never existed -- that seems like a reasonable position for transactions given that the timeframe is measured in months rather than minutes.

That should simplify the logic here as well, because I think you can skip this check entirely.